### PR TITLE
10) added new model and connected with it app levels

### PR DESCRIPTION
**Pull request by AntonPanchishko**

### CODE REVIEW

File: src/main/java/com/example/bankservicedemo/model/Transaction.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+@Entity(name = "transactions")
+@Data
+public class Transaction {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne
+    private Account accountFrom;
+    @OneToOne
+    private Account accountTo;
```
Comment (yurwing) [inline src/main/java/com/example/bankservicedemo/model/Transaction.java]: rename this columns for the table

**Review by yurwing — APPROVED** [review]

File: src/main/java/com/example/bankservicedemo/model/Account.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+public class Account {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(name = "account_number", unique = true, nullable = false)
+    private String accountNumber;
+    @Enumerated(EnumType.STRING)
+    @Column(nullable = false)
+    private Currency currency;
+    private double balance;
```
Comment (boroda4436) [inline src/main/java/com/example/bankservicedemo/model/Account.java]:
```suggestion
    private BigDecimal balance;
```
please use BigDecimal for money fields

File: src/main/java/com/example/bankservicedemo/model/Transaction.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne
+    @JoinColumn(name = "account_from")
+    private Account accountFrom;
+    @OneToOne
+    @JoinColumn(name = "account_to")
+    private Account accountTo;
+    @Min(0)
+    private double amount;
```
Comment (boroda4436) [inline src/main/java/com/example/bankservicedemo/model/Transaction.java]: BigDecimal

**Review by boroda4436 — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 4]
```diff
@@ -56,6 +56,15 @@
...
             <artifactId>lombok</artifactId>
         </dependency>
+        <dependency>
+            <groupId>javax.validation</groupId>
+            <artifactId>validation-api</artifactId>
+            <version>2.0.1.Final</version>
+        </dependency>
+        <dependency>
+            <groupId>com.jayway.jsonpath</groupId>
+            <artifactId>json-path</artifactId>
```
Comment (KseniiaMakarova) [inline pom.xml]: do you use this at this point?

File: src/main/java/com/example/bankservicedemo/model/Transaction.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+import javax.validation.constraints.Min;
+import lombok.Data;
+
+@Entity(name = "transactions")
+@Data
+public class Transaction {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bankservicedemo/model/Transaction.java]: why OneToOne?

File: src/main/java/com/example/bankservicedemo/model/Transaction.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    @JoinColumn(name = "account_from")
+    private Account accountFrom;
+    @OneToOne
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bankservicedemo/model/Transaction.java]: same

File: src/main/java/com/example/bankservicedemo/repository/AccountRepository.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+import com.example.bankservicedemo.model.Account;
+import java.util.List;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+
+public interface AccountRepository extends JpaRepository<Account, Long> {
+    @Query("SELECT a FROM accounts a JOIN FETCH a.user u WHERE u.phoneNumber = :phone")
+    List<Account> getAccountByPhone(String phone);
+
+    Account findByAccountNumber(String accountNumber);
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bankservicedemo/repository/AccountRepository.java]:
```suggestion
    Optional<Account> findByAccountNumber(String accountNumber);
```

File: src/main/java/com/example/bankservicedemo/repository/TransactionRepository.java [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+import com.example.bankservicedemo.model.Transaction;
+import java.util.List;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+
+public interface TransactionRepository extends JpaRepository<Transaction, Long> {
+    @Query("SELECT t FROM transactions t JOIN FETCH t.accountFrom af"
+            + " JOIN FETCH t.accountTo at"
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bankservicedemo/repository/TransactionRepository.java]: is it obligatory to use join fetch here?

File: src/main/java/com/example/bankservicedemo/repository/AccountRepository.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package com.example.bankservicedemo.repository;
+
+import com.example.bankservicedemo.model.Account;
+import java.util.List;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+
+public interface AccountRepository extends JpaRepository<Account, Long> {
+    @Query("SELECT a FROM accounts a JOIN FETCH a.user u WHERE u.phoneNumber = :phone")
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bankservicedemo/repository/AccountRepository.java:10]: same - try this
```suggestion
    @Query("SELECT a FROM accounts a WHERE a.user.phoneNumber = :phone")
```

File: src/main/java/com/example/bankservicedemo/service/AccountService.java [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+import java.util.List;
+
+public interface AccountService {
+    Account save(Account account);
+
+    List<Account> getAccountsByPhone(String phone);
+
+    Account getByAccountNumber(String accountNumber);
+
+    void blockAccount(String accountNumber);
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bankservicedemo/service/AccountService.java]: let's set `active` to false somewhere in controller and then just call `service.save()` - what do you think?

File: src/main/java/com/example/bankservicedemo/service/impl/TransactionServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+import org.springframework.stereotype.Service;
+
+@Service
+@AllArgsConstructor
+public class TransactionServiceImpl implements TransactionService {
+    private final TransactionRepository transactionRepository;
+    private final AccountService accountService;
+
+    @Override
+    public void transfer(String fromAccount, String toAccount, int amount) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bankservicedemo/service/impl/TransactionServiceImpl.java]:
```suggestion
    public void transfer(Account fromAccount, Account toAccount, int amount) {
```
and maybe we could move this to AccountService - how do you think?

File: src/main/java/com/example/bankservicedemo/service/impl/TransactionServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+        Transaction transactionOutcoming = new Transaction();
+        Account fromAccountByNumber = accountService.getByAccountNumber(fromAccount);
+        if (Double.parseDouble(String.valueOf(transactionOutcoming
+                .getAmount().subtract(BigDecimal.valueOf(amount)))) <= 0) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bankservicedemo/service/impl/TransactionServiceImpl.java]: what will be the value of `transactionOutcoming.getAmount()` here?
you created the empty object 2 lines ago

File: src/main/java/com/example/bankservicedemo/service/impl/TransactionServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+            throw new NoEnoughMoneyException("No enough money on balance");
+        }
+        Account toAccountNyNumber = accountService.getByAccountNumber(toAccount);
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bankservicedemo/service/impl/TransactionServiceImpl.java]:
```suggestion
        Account toAccountByNumber = accountService.getByAccountNumber(toAccount);
```

File: src/main/java/com/example/bankservicedemo/service/impl/TransactionServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+import org.springframework.data.domain.Sort;
+import org.springframework.stereotype.Service;
+
+@Service
+@AllArgsConstructor
+public class TransactionServiceImpl implements TransactionService {
+    private final TransactionRepository transactionRepository;
+    private final AccountService accountService;
+
+    @Override
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bankservicedemo/service/impl/TransactionServiceImpl.java:25]: let's use the `@Transactional` annotation to make this whole logic in this method be executed in a single transaction
see this article https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/transaction/annotation/Transactional.html

File: src/main/java/com/example/bankservicedemo/service/impl/TransactionServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+        accountService.save(toAccountNyNumber);
+        transactionRepository.save(transactionIncoming);
+    }
+
+    @Override
+    public List<Transaction> getAllByAccount(int page, int size, Account account) {
+        PageRequest sortedByDateAndId =
+                PageRequest.of(page, size, Sort.by("date").descending().and(Sort.by("id")));
+        return transactionRepository.findAllByAccount(account
+                .getAccountNumber(), sortedByDateAndId);
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bankservicedemo/service/impl/TransactionServiceImpl.java:59]:
```suggestion
        return transactionRepository.findAllByAccount(account, sortedByDateAndId);
```

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/bankservicedemo/controller/DataInsert.java [hunk 16]
```diff
@@ -12,15 +12,19 @@
...
+@RestController
+@RequestMapping("/insert")
 @AllArgsConstructor
 public class DataInsert {
     private final RoleService roleService;
     private final UserService userService;
     private final AccountService accountService;
 
+    @GetMapping
     @PostConstruct
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bankservicedemo/controller/DataInsert.java]: maybe let's remove @PostConstruct?

File: src/main/java/com/example/bankservicedemo/service/impl/AccountServiceImpl.java [hunk 17]
```diff
@@ -24,7 +25,11 @@ public Account save(Account account) {
...
     public Account getByAccountNumber(String accountNumber) {
-        return accountRepository.findByAccountNumber(accountNumber);
+        return accountRepository.findByAccountNumber(accountNumber).orElseThrow(()
+                -> {
+            throw new NoSuchEntityException("Can't find account"
+                    + " with such account number");
+        });
     }
 
     @Override
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bankservicedemo/service/impl/AccountServiceImpl.java]: let's remove blockAccount() method if we don't use it

File: src/main/java/com/example/bankservicedemo/service/impl/TransactionServiceImpl.java [hunk 18]
```diff
@@ -9,6 +9,7 @@
 import java.math.BigDecimal;
 import java.time.LocalDateTime;
 import java.util.List;
+import javax.transaction.Transactional;
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bankservicedemo/service/impl/TransactionServiceImpl.java]: use org.springframework.transaction.annotation.Transactional

File: src/main/java/com/example/bankservicedemo/repository/TransactionRepository.java [hunk 19]
```diff
@@ -7,9 +7,8 @@
 import org.springframework.data.jpa.repository.Query;
 
 public interface TransactionRepository extends JpaRepository<Transaction, Long> {
-    @Query("SELECT t FROM transactions t JOIN FETCH t.accountFrom af"
-            + " JOIN FETCH t.accountTo at"
-            + " WHERE af.accountNumber = :accountNumber OR"
-            + " at.accountNumber = :accountNumber")
+    @Query("SELECT t FROM transactions t "
+            + " WHERE t.accountTo = :accountNumber OR"
+            + " t.accountFrom = :accountNumber")
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bankservicedemo/repository/TransactionRepository.java]: can Account be equal to String?

File: src/main/java/com/example/bankservicedemo/repository/TransactionRepository.java [hunk 20]
```diff
@@ -7,9 +7,8 @@
...
     List<Transaction> findAllByAccount(String accountNumber, Pageable sortedByDateAndId);
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bankservicedemo/repository/TransactionRepository.java]: try `List<Transaction> findAllByAccount(Account account, Pageable sortedByDateAndId);`

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]
good job)

**Pull request merged**